In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [2]:
import glob
files = glob.glob('*.csv')

In [3]:
df1 = pd.read_csv('./p_data.csv', delimiter=';')
df1

Age      Sex          ID                 Race         Country  \
0      39.0     Male       49084                White   United-States   
1      50.0     Male       68553                White   United-States   
2      38.0     Male      143654                White   United-States   
3      53.0     Male       41510                Black   United-States   
4      28.0   Female      195025                Black            Cuba   
...     ...      ...         ...                  ...             ...   
48837  33.0     Male   597186815                White   United-States   
48838  39.0   Female   735617153                White   United-States   
48839  64.0     Male  1131807446                Black   United-States   
48840  38.0     Male   781422087                White   United-States   
48841  44.0     Male   161771296   Asian-Pac-Islander   United-States   

        Education  Education-Num       Marital Status     Relationship  
0       Bachelors           13.0        Never-married    Not-in-family  
1       Bachelors           13.0   Married-civ-spouse          Husband  
2         HS-grad            9.0             Divorced    Not-in-family  
3            11th            7.0   Married-civ-spouse          Husband  
4       Bachelors           13.0   Married-civ-spouse             Wife  
...           ...            ...                  ...              ...  
48837   Bachelors           13.0        Never-married        Own-child  
48838   Bachelors           13.0             Divorced    Not-in-family  
48839     HS-grad            9.0              Widowed   Other-relative  
48840   Bachelors           13.0   Married-civ-spouse          Husband  
48841   Bachelors           13.0             Divorced        Own-child  

[48842 rows x 9 columns]

In [4]:
data_list = []
for filename in files:
    data = pd.read_csv(filename, delimiter=';')
    data_list.append(data)
    
all_data = pd.concat(data_list)
all_data.head()

Age      Sex      ID    Race         Country   Education  Education-Num  \
0  39.0     Male   49084   White   United-States   Bachelors           13.0   
1  50.0     Male   68553   White   United-States   Bachelors           13.0   
2  38.0     Male  143654   White   United-States     HS-grad            9.0   
3  53.0     Male   41510   Black   United-States        11th            7.0   
4  28.0   Female  195025   Black            Cuba   Bachelors           13.0   

        Marital Status    Relationship Workclass  fnlwgt  Capital Gain  \
0        Never-married   Not-in-family       NaN     NaN           NaN   
1   Married-civ-spouse         Husband       NaN     NaN           NaN   
2             Divorced   Not-in-family       NaN     NaN           NaN   
3   Married-civ-spouse         Husband       NaN     NaN           NaN   
4   Married-civ-spouse            Wife       NaN     NaN           NaN   

   Capital Loss  Hours per week Target  
0           NaN             NaN    NaN  
1           NaN             NaN    NaN  
2           NaN             NaN    NaN  
3           NaN             NaN    NaN  
4           NaN             NaN    NaN

In [5]:
df2 = pd.read_csv('./w_data.csv', delimiter=';')
df2

Workclass    fnlwgt  Capital Gain  Capital Loss  \
0              State-gov   77516.0        2174.0           0.0   
1       Self-emp-not-inc   83311.0           0.0           0.0   
2                Private  215646.0           0.0           0.0   
3                Private  234721.0           0.0           0.0   
4                Private  338409.0           0.0           0.0   
...                  ...       ...           ...           ...   
48838            Private  215419.0           0.0           0.0   
48839                  ?  321403.0           0.0           0.0   
48840            Private  374983.0           0.0           0.0   
48841            Private   83891.0        5455.0           0.0   
48842       Self-emp-inc  182148.0           0.0           0.0   

       Hours per week          ID   Target  
0                40.0       49084    <=50K  
1                13.0       68553    <=50K  
2                40.0      143654    <=50K  
3                40.0       41510    <=50K  
4                40.0      195025    <=50K  
...               ...         ...      ...  
48838            36.0   735617153   <=50K.  
48839            40.0  1131807446   <=50K.  
48840            50.0   781422087   <=50K.  
48841            40.0   161771296   <=50K.  
48842            60.0  1344995623    >50K.  

[48843 rows x 7 columns]

In [6]:
df_merged_inner = pd.merge(df1,df2, on ='ID', how= 'inner')
df_merged_inner

Age      Sex          ID                 Race         Country  \
0      39.0     Male       49084                White   United-States   
1      50.0     Male       68553                White   United-States   
2      38.0     Male      143654                White   United-States   
3      53.0     Male       41510                Black   United-States   
4      28.0   Female      195025                Black            Cuba   
...     ...      ...         ...                  ...             ...   
48837  33.0     Male   597186815                White   United-States   
48838  39.0   Female   735617153                White   United-States   
48839  64.0     Male  1131807446                Black   United-States   
48840  38.0     Male   781422087                White   United-States   
48841  44.0     Male   161771296   Asian-Pac-Islander   United-States   

        Education  Education-Num       Marital Status     Relationship  \
0       Bachelors           13.0        Never-married    Not-in-family   
1       Bachelors           13.0   Married-civ-spouse          Husband   
2         HS-grad            9.0             Divorced    Not-in-family   
3            11th            7.0   Married-civ-spouse          Husband   
4       Bachelors           13.0   Married-civ-spouse             Wife   
...           ...            ...                  ...              ...   
48837   Bachelors           13.0        Never-married        Own-child   
48838   Bachelors           13.0             Divorced    Not-in-family   
48839     HS-grad            9.0              Widowed   Other-relative   
48840   Bachelors           13.0   Married-civ-spouse          Husband   
48841   Bachelors           13.0             Divorced        Own-child   

               Workclass    fnlwgt  Capital Gain  Capital Loss  \
0              State-gov   77516.0        2174.0           0.0   
1       Self-emp-not-inc   83311.0           0.0           0.0   
2                Private  215646.0           0.0           0.0   
3                Private  234721.0           0.0           0.0   
4                Private  338409.0           0.0           0.0   
...                  ...       ...           ...           ...   
48837            Private  245211.0           0.0           0.0   
48838            Private  215419.0           0.0           0.0   
48839                  ?  321403.0           0.0           0.0   
48840            Private  374983.0           0.0           0.0   
48841            Private   83891.0        5455.0           0.0   

       Hours per week   Target  
0                40.0    <=50K  
1                13.0    <=50K  
2                40.0    <=50K  
3                40.0    <=50K  
4                40.0    <=50K  
...               ...      ...  
48837            40.0   <=50K.  
48838            36.0   <=50K.  
48839            40.0   <=50K.  
48840            50.0   <=50K.  
48841            40.0   <=50K.  

[48842 rows x 15 columns]

In [7]:
df_merged_right = pd.merge(df1,df2, on ='ID', how= 'right')
df_merged_right

Age      Sex          ID                 Race         Country  \
0      39.0     Male       49084                White   United-States   
1      50.0     Male       68553                White   United-States   
2      38.0     Male      143654                White   United-States   
3      53.0     Male       41510                Black   United-States   
4      28.0   Female      195025                Black            Cuba   
...     ...      ...         ...                  ...             ...   
48838  39.0   Female   735617153                White   United-States   
48839  64.0     Male  1131807446                Black   United-States   
48840  38.0     Male   781422087                White   United-States   
48841  44.0     Male   161771296   Asian-Pac-Islander   United-States   
48842   NaN      NaN  1344995623                  NaN             NaN   

        Education  Education-Num       Marital Status     Relationship  \
0       Bachelors           13.0        Never-married    Not-in-family   
1       Bachelors           13.0   Married-civ-spouse          Husband   
2         HS-grad            9.0             Divorced    Not-in-family   
3            11th            7.0   Married-civ-spouse          Husband   
4       Bachelors           13.0   Married-civ-spouse             Wife   
...           ...            ...                  ...              ...   
48838   Bachelors           13.0             Divorced    Not-in-family   
48839     HS-grad            9.0              Widowed   Other-relative   
48840   Bachelors           13.0   Married-civ-spouse          Husband   
48841   Bachelors           13.0             Divorced        Own-child   
48842         NaN            NaN                  NaN              NaN   

               Workclass    fnlwgt  Capital Gain  Capital Loss  \
0              State-gov   77516.0        2174.0           0.0   
1       Self-emp-not-inc   83311.0           0.0           0.0   
2                Private  215646.0           0.0           0.0   
3                Private  234721.0           0.0           0.0   
4                Private  338409.0           0.0           0.0   
...                  ...       ...           ...           ...   
48838            Private  215419.0           0.0           0.0   
48839                  ?  321403.0           0.0           0.0   
48840            Private  374983.0           0.0           0.0   
48841            Private   83891.0        5455.0           0.0   
48842       Self-emp-inc  182148.0           0.0           0.0   

       Hours per week   Target  
0                40.0    <=50K  
1                13.0    <=50K  
2                40.0    <=50K  
3                40.0    <=50K  
4                40.0    <=50K  
...               ...      ...  
48838            36.0   <=50K.  
48839            40.0   <=50K.  
48840            50.0   <=50K.  
48841            40.0   <=50K.  
48842            60.0    >50K.  

[48843 rows x 15 columns]

In [8]:
df_merged_right.isnull().sum()

Age               2
Sex               2
ID                0
Race              2
Country           2
Education         2
Education-Num     2
Marital Status    2
Relationship      2
Workclass         1
fnlwgt            1
Capital Gain      1
Capital Loss      1
Hours per week    1
Target            1
dtype: int64

In [9]:
df_merged_right[df_merged_right['Sex'].isna()]

Age  Sex          ID Race Country Education  Education-Num  \
32561  NaN  NaN       37078  NaN     NaN       NaN            NaN   
48842  NaN  NaN  1344995623  NaN     NaN       NaN            NaN   

      Marital Status Relationship      Workclass    fnlwgt  Capital Gain  \
32561            NaN          NaN            NaN       NaN           NaN   
48842            NaN          NaN   Self-emp-inc  182148.0           0.0   

       Capital Loss  Hours per week  Target  
32561           NaN             NaN     NaN  
48842           0.0            60.0   >50K.

1) Tienen 48842 ID's en común. El único que no matchean es el 37078 ya que solo lo tiene el df2 (Por eso cuando hago el inner join no aparece y si cuando hago el right join)

In [10]:
#Es medio complicado por cómo lo encaraste. Otra opción simple sería usar el parámetro indicator de pd.merge
#En vez de concat
merged_indicator= pd.merge( df1,df2, on='ID', how='outer', indicator='origin')
merged_indicator[merged_indicator['origin'] != 'both']['ID']

48842    1344995623
Name: ID, dtype: int64

In [11]:
merged_indicator['origin']

0              both
1              both
2              both
3              both
4              both
            ...    
48838          both
48839          both
48840          both
48841          both
48842    right_only
Name: origin, Length: 48843, dtype: category
Categories (3, object): ['left_only', 'right_only', 'both']

In [12]:
df_merged_inner['Target'].value_counts()

 <=50K     24720
 <=50K.    12435
 >50K       7841
 >50K.      3845
Name: Target, dtype: int64

In [13]:
df_merged_inner['Target'].value_counts()

 <=50K     24720
 <=50K.    12435
 >50K       7841
 >50K.      3845
Name: Target, dtype: int64

In [14]:
df_merged_inner[df_merged_inner['Target'].isna()]

Age  Sex     ID Race Country Education  Education-Num Marital Status  \
32561  NaN  NaN  37078  NaN     NaN       NaN            NaN            NaN   

      Relationship Workclass  fnlwgt  Capital Gain  Capital Loss  \
32561          NaN       NaN     NaN           NaN           NaN   

       Hours per week Target  
32561             NaN    NaN

In [17]:
#Acá para resolver el probelma de los strings podrías primero eliminar el nan
#Y luego usar métodos de string a través de apply para evitar el uso de los fors 
#https://www.w3schools.com/python/python_ref_string.asp
df_merged_inner=df_merged_inner.dropna(subset=['Target'])
df_merged_inner["Target"] = df_merged_inner.Target.apply(lambda x: str(x).strip().replace(".",""))
#Siempre es conveniente usar apply en vez de for para cuando querés modificar los valores de una columna

In [19]:
df_merged_inner.Target.value_counts()

<=50K    37155
>50K     11686
Name: Target, dtype: int64

2) Listo

In [20]:
grouped_sex = df_merged_inner.groupby(["Sex", "Target"])
df_percentage_sex = grouped_sex.count()
gruped_percentage_sex = (100. * df_percentage_sex / df_percentage_sex.sum())
gruped_percentage_sex.rename(columns = {'Age':'Percentage'}, inplace = True)
gruped_percentage_sex = gruped_percentage_sex.drop(gruped_percentage_sex.columns[[1,2,3,4,5,6,7,8,9,10,11,12]], axis=1)
gruped_percentage_sex

Percentage
Sex     Target            
 Female <=50K    29.530517
        >50K      3.621957
 Male   <=50K    46.542864
        >50K     20.304662

In [21]:
df_merged_inner['Sex'].value_counts()

 Male      32649
 Female    16192
Name: Sex, dtype: int64

3.a) Podemos ver que el salario de los hombres >50K es 4X el de las mujeres en el mismo rango salarial. Tengamos en cuante que el dataset utilizado contiene el doble de cantidad de hombres.

In [22]:
grouped_race = df_merged_inner.groupby(["Race", "Target"])
df_percentage_race = grouped_race.count()
gruped_percentage_race = (100. * df_percentage_race/ df_percentage_race.sum())
gruped_percentage_race.rename(columns = {'Age':'Percentage'}, inplace = True)
gruped_percentage_race = gruped_percentage_race.drop(gruped_percentage_race.columns[[1,2,3,4,5,6,7,8,9,10,11,12]], axis=1)
gruped_percentage_race

Percentage
Race                Target            
 Amer-Indian-Eskimo <=50K     0.849696
                    >50K      0.112610
 Asian-Pac-Islander <=50K     2.272681
                    >50K      0.837411
 Black              <=50K     8.433488
                    >50K      1.158862
 Other              <=50K     0.728896
                    >50K      0.102373
 White              <=50K    63.788620
                    >50K     21.715362

In [23]:
df_merged_inner['Race'].value_counts()

 White                 41761
 Black                  4685
 Asian-Pac-Islander     1519
 Amer-Indian-Eskimo      470
 Other                   406
Name: Race, dtype: int64

3.b) Acá podemos apreciar que los hombres blancos con salarios >50K son el casi 22% frente a las otras etnias que rondan el 1% o menos. La cantidad de personas blancas en el dataset es también muy superior frente a las otras razas.

In [24]:
grouped_education = df_merged_inner.groupby(["Education", "Target"])
df_percentage_education = grouped_education.count()
gruped_percentage_education = (100. * df_percentage_education / df_percentage_education.sum())
gruped_percentage_education.rename(columns = {'Age':'Percentage'}, inplace = True)
gruped_percentage_education = gruped_percentage_education.drop(gruped_percentage_education.columns[[1,2,3,4,5,6,7,8,9,10,11,12]], axis=1)
gruped_percentage_education

Percentage
Education     Target            
 10th         <=50K     2.665793
              >50K      0.178129
 11th         <=50K     3.521631
              >50K      0.188366
 12th         <=50K     1.246903
              >50K      0.098278
 1st-4th      <=50K     0.489343
              >50K      0.016380
 5th-6th      <=50K     0.986876
              >50K      0.055281
 7th-8th      <=50K     1.828382
              >50K      0.126943
 9th          <=50K     1.463934
              >50K      0.083946
 Assoc-acdm   <=50K     2.432383
              >50K      0.845601
 Assoc-voc    <=50K     3.151041
              >50K      1.068774
 Bachelors    <=50K     9.647632
              >50K      6.781188
 Doctorate    <=50K     0.333736
              >50K      0.882455
 HS-grad      <=50K    27.192318
              >50K      5.124793
 Masters      <=50K     2.452857
              >50K      2.987244
 Preschool    <=50K     0.167892
              >50K      0.002047
 Prof-school  <=50K     0.444299
              >50K      1.263283
 Some-college <=50K    18.048361
              >50K      4.223910

In [25]:
df_merged_inner['Education'].value_counts()

 HS-grad         15784
 Some-college    10878
 Bachelors        8024
 Masters          2657
 Assoc-voc        2061
 11th             1812
 Assoc-acdm       1601
 10th             1389
 7th-8th           955
 Prof-school       834
 9th               756
 12th              657
 Doctorate         594
 5th-6th           509
 1st-4th           247
 Preschool          83
Name: Education, dtype: int64

3.c) Los salarios mas altos (>50K) se dan en los que tienen Education Bachelors, Master y HS-Grad.

In [27]:
#Muy bien hecho todo el análisis y las agrupaciones. Me parece que la consigna preguntaba solo sobre el
#subconjunto de >50K. Así que podrías reducir todo de esta forma:
df_merged_inner_50k= df_merged_inner[df_merged_inner['Target'] == '>50K']

columns=['Sex', 'Education', 'Race']
for col in columns:
    print('#############')
    print(f"Mayores a 50K segun {col}")
    print('#############')
    print(df_merged_inner_50k[col].value_counts(normalize=True, ascending=False).round(2))
    print('-------------')

#############
Mayores a 50K segun Sex
#############
 Male      0.85
 Female    0.15
Name: Sex, dtype: float64
-------------
#############
Mayores a 50K segun Education
#############
 Bachelors       0.28
 HS-grad         0.21
 Some-college    0.18
 Masters         0.12
 Prof-school     0.05
 Assoc-voc       0.04
 Doctorate       0.04
 Assoc-acdm      0.04
 11th            0.01
 10th            0.01
 7th-8th         0.01
 12th            0.00
 9th             0.00
 5th-6th         0.00
 1st-4th         0.00
 Preschool       0.00
Name: Education, dtype: float64
-------------
#############
Mayores a 50K segun Race
#############
 White                 0.91
 Black                 0.05
 Asian-Pac-Islander    0.03
 Amer-Indian-Eskimo    0.00
 Other                 0.00
Name: Race, dtype: float64
-------------


In [ ]:
#Ahi tenés cómo se distribuye según cada categoría el segmento de >50K. Condice con lo que decis
#pero los números son diferentes porque cambia la base. En vez del total es el total de >50K